rust/crd/src/lib.rs

```diff
@@ -804,13 +804,22 @@ impl HdfsCluster {
         Ok(result)
     }
 
-    pub fn is_upgrading(&self) -> bool {
-        self.status
-            .as_ref()
-            .and_then(|status| status.deployed_product_version.as_deref())
-            .map_or(false, |deployed_version| {
-                deployed_version != self.spec.image.product_version()
-            })
+    pub fn upgrade_state(&self) -> Option<UpgradeState> {
+        let status = self.status.as_ref()?;
+        let requested_version = self.spec.image.product_version();
+
+        if requested_version != status.deployed_product_version.as_deref()? {
+            // If we're requesting a different version than what is deployed, assume that we're upgrading.
+            // Could also be a downgrade to an older version, but we don't support downgrades after upgrade finalization.
```

Comment:
Basically, we support downgrading back *during* the upgrade, but only until you `rollingUpgrade -finalize`. After that you're "stuck". We also don't support something like "start at 3.3.4, upgrade to 3.3.6, abort and downgrade to 3.3.5", you would have to downgrade to 3.3.4 and *then* do a separate upgrade to 3.3.5.

Reply:
> We also don't support something like "start at 3.3.4, upgrade to 3.3.6, abort and downgrade to 3.3.5", you would have to downgrade to 3.3.4 and then do a separate upgrade to 3.3.5.

Is there a guardrail for that case? Is it detected in the current code, or we just document what we do/don't support.

_I thought it would be quicker to ask than to rerun through the PR and guess_.

Reply:
There isn't currently, but I think it'd make sense to add one.

Reply:
2a25ff4071412a48843b0fa108f0029677863511